toreview/PeerNode.java

```diff
@@ -0,0 +1,94 @@
+package com.gruppo1.distributednetworkmanager;
+
+import androidx.annotation.NonNull;
+
+import com.dezen.riccardo.smshandler.Peer;
+import com.dezen.riccardo.smshandler.SMSPeer;
+
+/**
+ * @author Niccolo' Turcato
+ * Class that represents the Peer DistributedNetworkNode of DistributedNetwork
+ * Address is a Bitset of fixed length
+ */
+public class PeerNode extends Peer<BinarySet> implements Node<BinarySet> {
+    private BinarySet binaryKey;
+    private SMSPeer physicalPeer;
+
+    /**
+     * Constructor that initializes the key with the given value
+     *
+     * @param buildingKey given value for the node's Key
+     */
+    public PeerNode(@NonNull BinarySet buildingKey) {
+        binaryKey = (BinarySet) buildingKey.clone();
+    }
+
+    public void setPhysicalPeer(SMSPeer peer) {
+        if (peer != null && peer.isValid())
+            physicalPeer = peer;
+    }
+
+    public SMSPeer getPhysicalPeer() {
+        return physicalPeer;
+    }
+
+
+    /**
+     * @return binary address of the node
+     */
+    public BinarySet getAddress() {
+        return (BinarySet) binaryKey.clone();
+    }
+
+    public BinarySet getKey() {
+        return (BinarySet) binaryKey.clone();
+    }
+

```

Comment:
questi due metodi fanno la stessa cosa! Rivedi l'interfaccia `Node`

Reply:
Inoltre cloni sempre gli oggetti per non restituire il riferimento. Perché? Puoi definire una specifica in cui esplicitamente dichiari che viene restituito il riferimento. Sarà compito di chi usa questo riferimento stare attento a non modificare l'oggetto riferito. Se deve fare modifiche lo clona direttamente lui, inutile farlo sempre se non è necessario.

Reply:
Hai ragione, sono uguali ma a livello strutturale non siamo riusciti ancora a decidere come evitare questa problematica, quindi è rimasta così.
Clono sempre la chiave perché i metodi di algebra booleana di BitSet sono distruttivi, quindi se qualcuno estraesse la chiave come riferimento a quella contenuta in uno di questi oggetti, basta un errore che va a modificare la chiave dell'oggetto.
Vero che posso sbolognare il compito di stare attento a chi usa l'oggetto, ma visto che bitset è clonable, e ci vuole poco poco sforzo a non passare il riferimento, ho preferito farlo. Evito altre rogne.

Reply:
Questi due metodi servono a restituire la chiave del nodo, che sarebbe il suo ID, giusto?
Il fatto che tu gli dia il nome di KEY o di ADDRESS o di ID cambia poco, credo che il consiglio migliore sia quello di decidere a livello di gruppo una nomenclatura che tutti utilizzino per rappresentare il valore dell'ID, che alla fine quello è, indifferentemente dallo scopo per cui lo estrai dal nodo

Reply:
Sì, il fatto di aver definito a monte due nomi diversi è stata una poor choice, parlavo anche più sopra del fatto che mi è stato un po' impedito di cambiare i nomi ai metodi delle interfacce per problemi con la repo (cosa che abbiamo risolto solo poco fa e non ho fatto in tempo ad agire sul codice.